### PR TITLE
Demo (Training) Updates

### PR DESCRIPTION
**Pull request by michael-grace**
**What's Changed**

- You can now select what your demo is for
- You can now integrate a link to indicate your demo is virtual (will email trainees the link  too)
- You can now edit demos
- You can sign up to training waiting lists and will be informed when slots become available
- Menu changes to see available training sessions better if you're already trained
- Tick box to easily mark all attendants of a training session as trained, rather than finding their individual profiles
- Multiple demos can now be simultaneously, and at the same time as shows
- New database structure for demos

**Not Done**
- Changed the iframe of the schedule
- Demos can currently be edited by anybody who can train, change this so only the demoer or people with appropriate positions.
- Be able to change the demoer for a session
- Allow demoers to view waiting lists, to see whats popular
- Getting all training statuses and working out which ones a user can be awarded is a slow process. Both listing waiting lists and other user's profiles take a little bit of time to load because of it.

### CODE REVIEW

**Comment by URY-GitHub** [comment]
Thank you for this pull request, @michael-grace!

@SplashyFish14, would you mind reviewing this PR, please? Feel free to ask anyone for help if you're not quite sure what you're doing.

**Comment by michael-grace** [comment]
membership DB changes not done yet

File: src/Menus/Scheduler.json [hunk 1]
```diff
@@ -35,6 +35,16 @@
                 "title": "Manage Terms",
                 "url": "module=Scheduler,action=listTerms",
                 "description": "Seasons are applied to in relation to Terms. Manage Terms here."
+            },
```
Comment (markspolakovs) [inline src/Menus/Scheduler.json:38]: Worth moving these into a separate module? Scheduler is getting quite chonk

File: src/Classes/ServiceAPI/MyRadio_Demo.php [hunk 2]
```diff
@@ -18,45 +20,127 @@
...
+        return $this->demo_id;
+    }
+
+    public static function registerDemo($time, $link = null, $training_type)
+    {
+        if ($time == null || $training_type == null) {
+            throw new MyRadioException("A training demo must have a time and training date.", 400);
+        } else {
+            self::initDB();
+            date_default_timezone_set('UTC');
```
Comment (markspolakovs) [inline src/Classes/ServiceAPI/MyRadio_Demo.php:63]: aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa

File: src/Classes/ServiceAPI/MyRadio_Demo.php [hunk 3]
```diff
@@ -18,45 +20,127 @@
...
+    }
 
+    public function getID()
+    {
+        return $this->demo_id;
+    }
+
+    public static function registerDemo($time, $link = null, $training_type)
+    {
+        if ($time == null || $training_type == null) {
```
Comment (markspolakovs) [inline src/Classes/ServiceAPI/MyRadio_Demo.php:59]: Worth validating type as well, not just existence?

File: src/Classes/ServiceAPI/MyRadio_Demo.php [hunk 4]
```diff
@@ -18,45 +20,127 @@
...
+            // Let people waiting for this know
+            $training = MyRadio_TrainingStatus::getInstance($training_type);
+            $waiters = self::trainingWaitingList($training_type);
+            foreach ($waiters as $waiter) {
+                $user = MyRadio_User::getInstance($waiter['memberid']);
+                MyRadioEmail::sendEmailToUser(
+                    $user,
+                    "Available Training Session",
+                    "Hi " . $user->getFName() . ","
+                        . "\r\n\r\n A session to get you " . $training->getTitle() . " has opened up. Check it out on MyRadio.\r\n\r\n"
```
Comment (markspolakovs) [inline src/Classes/ServiceAPI/MyRadio_Demo.php:81]: Link?

File: src/Classes/ServiceAPI/MyRadio_Demo.php [hunk 5]
```diff
@@ -18,45 +20,127 @@
...
+                );
+            }
+        }
         return true;
     }
 
+    public function editDemo($time, $link = null, $training_type)
+    {
+
+        // TODO, Only edit your training demos, or any if you have perms
```
Comment (markspolakovs) [inline src/Classes/ServiceAPI/MyRadio_Demo.php:92]: This seems like a pretty big TODO

File: src/Classes/ServiceAPI/MyRadio_Demo.php [hunk 6]
```diff
@@ -18,45 +20,127 @@
...
+                $attendees[] = $this->getDemoer();
+                foreach ($attendees as $attendee) {
+                    MyRadioEmail::sendEmailToUser(
+                        $attendee,
+                        "Updated Training Session",
+                        "Hi " . $attendee->getFName()
+                            . "\r\n\r\n There's been a change to your training session on " . $this->demo_time
+                            . ".\r\n\r\n"
+                            . ($time != $this->demo_time ? "It is now at " . CoreUtils::happyTime($time) . ".\r\n\r\n" : "")
+                            . (($link == null && $this->demo_link != null) ? "It is now in person at our studios in Vanbrugh College.\r\n\r\n" : (($link != null && $this->demo_link == null) ? "It is now online and will be hosted at " . $link . "\r\n\r\n" : (($link != null) ? "The session is now at " . $link . "\r\n\r\n" : "\r\n\r\n")))
```
Comment (markspolakovs) [inline src/Classes/ServiceAPI/MyRadio_Demo.php:116]: Eeeeeee, this is quite hard to read. Would you mind breaking it into variables?

File: src/Classes/ServiceAPI/MyRadio_Demo.php [hunk 7]
```diff
@@ -65,48 +149,89 @@ public static function getForm()
...
-            $str .= ', '.MyRadio_User::getInstance($r[1])->getName();
+            $str .= ', ' . MyRadio_User::getInstance($r[1])->getName();
         }
 
         return $str;
     }
 
-    public static function attendingDemoCount($demoid)
+    // Lets fix the above with this...grrr....
+    public function myRadioUsersAttendingDemo()
```
Comment (markspolakovs) [inline src/Classes/ServiceAPI/MyRadio_Demo.php:216]: No. Bad Michael.

File: src/Classes/ServiceAPI/MyRadio_Demo.php [hunk 8]
```diff
@@ -137,105 +264,194 @@ public static function listDemos()
...
-            .' Training'
+                . "Just to confirm that you have left the training session at " . $this->getDemoTime() . ". If this was accidental, simply rejoin. "
+                . "Meanwhile, you've been added to the waiting list, and we'll let you know if a session becomes available."
+                . "\r\n\r\nThanks!\r\n"
+                . Config::$long_name
+                . ' Training'
         );
 
+        // Add to waiting list
+        self::joinWaitingList($this->presenterstatusid);
```
Comment (markspolakovs) [inline src/Classes/ServiceAPI/MyRadio_Demo.php:353]: Is it a good idea to do this automatically? I'm not convinced

File: src/Controllers/Scheduler/createDemo.php [hunk 9]
```diff
@@ -8,14 +8,26 @@
...
     $demoinfo = MyRadio_Demo::getForm()->readValues();
-
-    MyRadio_Demo::registerDemo($demoinfo['demo-datetime']);
-
+    if ($demoinfo['id']){
+        // Update a Demo
+        MyRadio_Demo::getInstance($demoinfo['id'])->editDemo($demoinfo['demo-datetime'], $demoinfo['demo-link'], $demoinfo['demo-training-type']);
+    }else{
+        // Create a New Demo
+    MyRadio_Demo::registerDemo($demoinfo['demo-datetime'], $demoinfo['demo-link'], $demoinfo['demo-training-type']);
```
Comment (markspolakovs) [inline src/Controllers/Scheduler/createDemo.php:16]: Consistency dashes v underscores

**Review by markspolakovs — CHANGES_REQUESTED** [review]

**Comment by SplashyFish14** [comment]
@michael-grace is this still a thing?

**Comment by michael-grace** [comment]
TBF, I need to finish it. Maybe I'll do that at some point soon... 😂

**Comment by mstratford** [comment]
@michael-grace poke

File: src/Menus/Scheduler.json (hunk 1 above)
Comment (michael-grace) [reply src/Menus/Scheduler.json:38]: ye...maybe. but not now (mainly cause i can't really be bothered to) because all changing pemissions and fun :)

**Comment by michael-grace** [comment]
> membership DB changes not done yet

i'm going to guess this still hasn't been done

File: src/Classes/ServiceAPI/MyRadio_Demo.php [hunk 10]
```diff
@@ -343,14 +361,13 @@ public function leave()
             'Training Cancellation',
             'Hi ' . $attendee->getFName() . ",\r\n\r\n"
                 . "Just to confirm that you have left the training session at " . $this->getDemoTime() . ". If this was accidental, simply rejoin. "
-                . "Meanwhile, you've been added to the waiting list, and we'll let you know if a session becomes available."
+                . "Meanwhile, you've can join the waiting list, and we'll let you know if a session becomes available. "
```
Comment (markspolakovs) [inline src/Classes/ServiceAPI/MyRadio_Demo.php]: Err, hang on - why are you telling them to join the waiting list when they've just left a sesh?

Comment (michael-grace) [reply src/Classes/ServiceAPI/MyRadio_Demo.php]: If they've left cause they can't make the time, doesn't necessarily mean they don't want the training.

Comment (markspolakovs) [reply src/Classes/ServiceAPI/MyRadio_Demo.php]: > you've can join the waiting list

Comment (michael-grace) [reply src/Classes/ServiceAPI/MyRadio_Demo.php]: https://github.com/UniversityRadioYork/MyRadio/pull/934/commits/3aee163c4db5c963851d5356654e4bc1ac2ef90c#diff-2532bcdab17654f155a88878e22e5c7665fa6ffcf2e813d4a306a5473801cc9bR368

**Comment by michael-grace** [comment]
TODO.
1. Add SQL patch to DB
2. Sort out permissions for new module

**Comment by Joseph-S12** [comment]
It's Michael's fault if this breaks.

**Pull request merged**